### ISSUE TITLE
Installation issue. 

### ISSUE THREAD

**Issue by noodlehair-git** [oscarlab/graphene#2095]
I was trying to re-install Graphene with the latest update. I followed the steps from https://graphene.readthedocs.io/en/latest/building.html# (chose step 2b for now). When I did ```make SGX=1```, I got the following error: 

```
make[3]: Entering directory '/home/arushiarora/sgx/graphene/Pal/src/host/Linux-SGX'
./link-intel-driver.py < gsgx.h.in > gsgx.h
ISGX_DRIVER_PATH environment variable is undefined. You can define
ISGX_DRIVER_PATH="" to use the default in-kernel driver's C header.
Makefile:150: recipe for target 'gsgx.h' failed
make[3]: *** [gsgx.h] Error 1
make[3]: Leaving directory '/home/arushiarora/sgx/graphene/Pal/src/host/Linux-SGX'
Makefile:101: recipe for target 'host_lib_recurse' failed
make[2]: *** [host_lib_recurse] Error 2
make[2]: Leaving directory '/home/arushiarora/sgx/graphene/Pal/src'
Makefile:7: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/home/arushiarora/sgx/graphene/Pal'
Makefile:13: recipe for target 'all' failed
make: *** [all] Error 2
```

I then set the value of ```ISGX_DRIVER_PATH="/path/to/my/sgx/linux-sgx-driver"```. 

```
lsmod | grep sgx
```
gives:
```
gsgx                   16384  0
isgx                   49152  2
```
and

```
ps ax | grep [a]esm_service
```
gives:

```
17289 ?        Ssl    0:00 /opt/intel/sgx-aesm-service/aesm/aesm_service
```

(Intel SDK is working fine as I could run LocalAttestation Example in HW mode).

But ```make SGX=1``` still throws the following error: 

```
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18:0,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c: In function ‘open_sgx_driver’:
sgx_framework.c:19:49: error: ‘GSGX_FILE’ undeclared (first use in this function); did you mean ‘STT_FILE’?
         g_gsgx_device = INLINE_SYSCALL(open, 3, GSGX_FILE, O_RDWR | O_CLOEXEC, 0);
                                                 ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:166:26: note: in definition of macro ‘LOAD_ARGS_1’
     long __arg1 = (long)(a1); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:183:5: note: in expansion of macro ‘LOAD_ARGS_2’
     LOAD_ARGS_2(a1, a2)
     ^~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:19:25: note: in expansion of macro ‘INLINE_SYSCALL’
         g_gsgx_device = INLINE_SYSCALL(open, 3, GSGX_FILE, O_RDWR | O_CLOEXEC, 0);
                         ^~~~~~~~~~~~~~
sgx_framework.c:19:49: note: each undeclared identifier is reported only once for each function it appears in
         g_gsgx_device = INLINE_SYSCALL(open, 3, GSGX_FILE, O_RDWR | O_CLOEXEC, 0);
                                                 ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:166:26: note: in definition of macro ‘LOAD_ARGS_1’
     long __arg1 = (long)(a1); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:183:5: note: in expansion of macro ‘LOAD_ARGS_2’
     LOAD_ARGS_2(a1, a2)
     ^~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:19:25: note: in expansion of macro ‘INLINE_SYSCALL’
         g_gsgx_device = INLINE_SYSCALL(open, 3, GSGX_FILE, O_RDWR | O_CLOEXEC, 0);
                         ^~~~~~~~~~~~~~
In file included from sgx_framework.c:5:0:
sgx_framework.c:25:96: error: expected ‘)’ before ‘GSGX_FILE’
 he Graphene SGX module is loaded, check permissions on the device " GSGX_FILE ", as we cannot open this file.\n\n");
                                                                     ^
pal_linux.h:340:24: note: in definition of macro ‘SGX_DBG’
             pal_printf(fmt);     \
                        ^~~
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18:0,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c:30:45: error: ‘ISGX_FILE’ undeclared (first use in this function); did you mean ‘GSGX_FILE’?
     g_isgx_device = INLINE_SYSCALL(open, 3, ISGX_FILE, O_RDWR | O_CLOEXEC, 0);
                                             ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:166:26: note: in definition of macro ‘LOAD_ARGS_1’
     long __arg1 = (long)(a1); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:183:5: note: in expansion of macro ‘LOAD_ARGS_2’
     LOAD_ARGS_2(a1, a2)
     ^~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:30:21: note: in expansion of macro ‘INLINE_SYSCALL’
     g_isgx_device = INLINE_SYSCALL(open, 3, ISGX_FILE, O_RDWR | O_CLOEXEC, 0);
                     ^~~~~~~~~~~~~~
In file included from sgx_framework.c:5:0:
sgx_framework.c:32:46: error: expected ‘)’ before ‘ISGX_FILE’
         SGX_DBG(DBG_E, "Cannot open device " ISGX_FILE ". "
                                              ^
pal_linux.h:340:24: note: in definition of macro ‘SGX_DBG’
             pal_printf(fmt);     \
                        ^~~
sgx_framework.c: In function ‘create_enclave’:
sgx_framework.c:178:12: error: variable ‘param’ has initializer but incomplete type
     struct sgx_enclave_create param = {
            ^~~~~~~~~~~~~~~~~~
sgx_framework.c:179:10: error: ‘struct sgx_enclave_create’ has no member named src’
         .src = (uint64_t)secs,
          ^~~
sgx_framework.c:179:16: warning: excess elements in struct initializer
         .src = (uint64_t)secs,
                ^
sgx_framework.c:179:16: note: (near initialization for ‘param’)
sgx_framework.c:178:31: error: storage size of ‘param’ isn’t known
     struct sgx_enclave_create param = {
                               ^~~~~
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18:0,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c:181:55: error: ‘SGX_IOC_ENCLAVE_CREATE’ undeclared (first use in this function); did you mean ‘DEFAULT_ENCLAVE_BASE’?
     int ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_CREATE, &param);
                                                       ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:174:26: note: in definition of macro ‘LOAD_ARGS_2’
     long __arg2 = (long)(a2); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:181:15: note: in expansion of macro ‘INLINE_SYSCALL’
     int ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_CREATE, &param);
               ^~~~~~~~~~~~~~
sgx_framework.c:178:31: warning: unused variable ‘param’ [-Wunused-variable]
     struct sgx_enclave_create param = {
                               ^~~~~
sgx_framework.c: In function ‘add_pages_to_enclave’:
sgx_framework.c:335:12: error: variable ‘param’ has initializer but incomplete type
     struct sgx_enclave_add_page param = {
            ^~~~~~~~~~~~~~~~~~~~
sgx_framework.c:336:10: error: ‘struct sgx_enclave_add_page’ has no member named ‘addr’
         .addr    = (uint64_t)addr,
          ^~~~
sgx_framework.c:336:20: warning: excess elements in struct initializer
         .addr    = (uint64_t)addr,
                    ^
sgx_framework.c:336:20: note: (near initialization for ‘param’)
sgx_framework.c:337:10: error: ‘struct sgx_enclave_add_page’ has no member named ‘src’
         .src     = (uint64_t)(user_addr ?: g_zero_pages),
          ^~~
sgx_framework.c:337:20: warning: excess elements in struct initializer
         .src     = (uint64_t)(user_addr ?: g_zero_pages),
                    ^
sgx_framework.c:337:20: note: (near initialization for ‘param’)
sgx_framework.c:338:10: error: ‘struct sgx_enclave_add_page’ has no member named ‘secinfo’
         .secinfo = (uint64_t)&secinfo,
          ^~~~~~~
sgx_framework.c:338:20: warning: excess elements in struct initializer
         .secinfo = (uint64_t)&secinfo,
                    ^
sgx_framework.c:338:20: note: (near initialization for ‘param’)
sgx_framework.c:339:10: error: ‘struct sgx_enclave_add_page’ has no member named ‘mrmask’
         .mrmask  = skip_eextend ? 0 : (uint16_t)-1,
          ^~~~~~
sgx_framework.c:339:20: warning: excess elements in struct initializer
         .mrmask  = skip_eextend ? 0 : (uint16_t)-1,
                    ^~~~~~~~~~~~
sgx_framework.c:339:20: note: (near initialization for ‘param’)
sgx_framework.c:335:33: error: storage size of ‘param’ isn’t known
     struct sgx_enclave_add_page param = {
                                 ^~~~~
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18:0,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c:344:55: error: ‘SGX_IOC_ENCLAVE_ADD_PAGE’ undeclared (first use in this function)
         ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_ADD_PAGE, &param);
                                                       ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:174:26: note: in definition of macro ‘LOAD_ARGS_2’
     long __arg2 = (long)(a2); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:344:15: note: in expansion of macro ‘INLINE_SYSCALL’
         ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_ADD_PAGE, &param);
               ^~~~~~~~~~~~~~
sgx_framework.c:335:33: warning: unused variable ‘param’ [-Wunused-variable]
     struct sgx_enclave_add_page param = {
                                 ^~~~~
sgx_framework.c: In function ‘init_enclave’:
sgx_framework.c:380:12: error: variable ‘param’ has initializer but incomplete type
     struct sgx_enclave_init param = {
            ^~~~~~~~~~~~~~~~
sgx_framework.c:382:10: error: ‘struct sgx_enclave_init’ has no member named ‘addr’
         .addr = enclave_valid_addr,
          ^~~~
sgx_framework.c:382:17: warning: excess elements in struct initializer
         .addr = enclave_valid_addr,
                 ^~~~~~~~~~~~~~~~~~
sgx_framework.c:382:17: note: (near initialization for ‘param’)
sgx_framework.c:384:10: error: ‘struct sgx_enclave_init’ has no member named ‘sigstruct’
         .sigstruct = (uint64_t)sigstruct,
          ^~~~~~~~~
sgx_framework.c:384:22: warning: excess elements in struct initializer
         .sigstruct = (uint64_t)sigstruct,
                      ^
sgx_framework.c:384:22: note: (near initialization for ‘param’)
sgx_framework.c:386:10: error: ‘struct sgx_enclave_init’ has no member named ‘einittoken’
         .einittoken = (uint64_t)token,
          ^~~~~~~~~~
sgx_framework.c:386:23: warning: excess elements in struct initializer
         .einittoken = (uint64_t)token,
                       ^
sgx_framework.c:386:23: note: (near initialization for ‘param’)
sgx_framework.c:380:29: error: storage size of ‘param’ isn’t known
     struct sgx_enclave_init param = {
                             ^~~~~
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18:0,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c:389:55: error: ‘SGX_IOC_ENCLAVE_INIT’ undeclared (first use in this function); did you mean ‘ATOMIC_FLAG_INIT’?
     int ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_INIT, &param);
                                                       ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:174:26: note: in definition of macro ‘LOAD_ARGS_2’
     long __arg2 = (long)(a2); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:389:15: note: in expansion of macro ‘INLINE_SYSCALL’
     int ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_INIT, &param);
               ^~~~~~~~~~~~~~
sgx_framework.c:399:18: error: ‘SGX_INVALID_SIG_STRUCT’ undeclared (first use in this function); did you mean ‘SGX_LINKABLE_SIGNATURE’?
             case SGX_INVALID_SIG_STRUCT:
                  ^~~~~~~~~~~~~~~~~~~~~~
                  SGX_LINKABLE_SIGNATURE
sgx_framework.c:402:18: error: ‘SGX_INVALID_ATTRIBUTE’ undeclared (first use in this function); did you mean ‘SGX_INVALID_SIG_STRUCT’?
             case SGX_INVALID_ATTRIBUTE:
                  ^~~~~~~~~~~~~~~~~~~~~
                  SGX_INVALID_SIG_STRUCT
sgx_framework.c:405:18: error: ‘SGX_INVALID_MEASUREMENT’ undeclared (first use in this function); did you mean ‘SGX_INVALID_ATTRIBUTE’?
             case SGX_INVALID_MEASUREMENT:
                  ^~~~~~~~~~~~~~~~~~~~~~~
                  SGX_INVALID_ATTRIBUTE
sgx_framework.c:408:18: error: ‘SGX_INVALID_SIGNATURE’ undeclared (first use in this function); did you mean ‘SGX_INVALID_SIG_STRUCT’?
             case SGX_INVALID_SIGNATURE:
                  ^~~~~~~~~~~~~~~~~~~~~
                  SGX_INVALID_SIG_STRUCT
sgx_framework.c:411:18: error: ‘SGX_INVALID_LICENSE’ undeclared (first use in this function); did you mean ‘SGX_INVALID_SIGNATURE’?
             case SGX_INVALID_LICENSE:
                  ^~~~~~~~~~~~~~~~~~~
                  SGX_INVALID_SIGNATURE
sgx_framework.c:414:18: error: ‘SGX_INVALID_CPUSVN’ undeclared (first use in this function); did you mean ‘SGX_INVALID_LICENSE’?
             case SGX_INVALID_CPUSVN:
                  ^~~~~~~~~~~~~~~~~~
                  SGX_INVALID_LICENSE
sgx_framework.c:380:29: warning: unused variable ‘param’ [-Wunused-variable]
     struct sgx_enclave_init param = {
                             ^~~~~
Makefile:117: recipe for target 'sgx_framework.o' failed
make[3]: *** [sgx_framework.o] Error 1
make[3]: Leaving directory '/home/arushiarora/sgx/graphene/Pal/src/host/Linux-SGX'
Makefile:101: recipe for target 'host_lib_recurse' failed
make[2]: *** [host_lib_recurse] Error 2
make[2]: Leaving directory '/home/arushiarora/sgx/graphene/Pal/src'
Makefile:7: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/home/arushiarora/sgx/graphene/Pal'
Makefile:13: recipe for target 'all' failed
make: *** [all] Error 2
arushiarora@NUC10i7FNH:~/sgx/graphene$ make SGX=1 DEBUG=1
make -C Scripts all
make[1]: Entering directory '/home/arushiarora/sgx/graphene/Scripts'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/arushiarora/sgx/graphene/Scripts'
make -C Pal all
make[1]: Entering directory '/home/arushiarora/sgx/graphene/Pal'
make -C src
make[2]: Entering directory '/home/arushiarora/sgx/graphene/Pal/src'
make -C ../lib target=/home/arushiarora/sgx/graphene/Pal/src/host/Linux-SGX/.lib/
make[3]: Entering directory '/home/arushiarora/sgx/graphene/Pal/lib'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/home/arushiarora/sgx/graphene/Pal/lib'
make[3]: Entering directory '/home/arushiarora/sgx/graphene/Pal/src/host/Linux-SGX'
  [ sgx_framework.o ]
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18:0,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c: In function ‘open_sgx_driver’:
sgx_framework.c:19:49: error: ‘GSGX_FILE’ undeclared (first use in this function); did you mean ‘STT_FILE’?
         g_gsgx_device = INLINE_SYSCALL(open, 3, GSGX_FILE, O_RDWR | O_CLOEXEC, 0);
                                                 ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:166:26: note: in definition of macro ‘LOAD_ARGS_1’
     long __arg1 = (long)(a1); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:183:5: note: in expansion of macro ‘LOAD_ARGS_2’
     LOAD_ARGS_2(a1, a2)
     ^~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:19:25: note: in expansion of macro ‘INLINE_SYSCALL’
         g_gsgx_device = INLINE_SYSCALL(open, 3, GSGX_FILE, O_RDWR | O_CLOEXEC, 0);
                         ^~~~~~~~~~~~~~
sgx_framework.c:19:49: note: each undeclared identifier is reported only once for each function it appears in
         g_gsgx_device = INLINE_SYSCALL(open, 3, GSGX_FILE, O_RDWR | O_CLOEXEC, 0);
                                                 ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:166:26: note: in definition of macro ‘LOAD_ARGS_1’
     long __arg1 = (long)(a1); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:183:5: note: in expansion of macro ‘LOAD_ARGS_2’
     LOAD_ARGS_2(a1, a2)
     ^~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:19:25: note: in expansion of macro ‘INLINE_SYSCALL’
         g_gsgx_device = INLINE_SYSCALL(open, 3, GSGX_FILE, O_RDWR | O_CLOEXEC, 0);
                         ^~~~~~~~~~~~~~
In file included from sgx_framework.c:5:0:
sgx_framework.c:25:96: error: expected ‘)’ before ‘GSGX_FILE’
 he Graphene SGX module is loaded, check permissions on the device " GSGX_FILE ", as we cannot open this file.\n\n");
                                                                     ^
pal_linux.h:340:24: note: in definition of macro ‘SGX_DBG’
             pal_printf(fmt);     \
                        ^~~
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18:0,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c:30:45: error: ‘ISGX_FILE’ undeclared (first use in this function); did you mean ‘GSGX_FILE’?
     g_isgx_device = INLINE_SYSCALL(open, 3, ISGX_FILE, O_RDWR | O_CLOEXEC, 0);
                                             ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:166:26: note: in definition of macro ‘LOAD_ARGS_1’
     long __arg1 = (long)(a1); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:183:5: note: in expansion of macro ‘LOAD_ARGS_2’
     LOAD_ARGS_2(a1, a2)
     ^~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:30:21: note: in expansion of macro ‘INLINE_SYSCALL’
     g_isgx_device = INLINE_SYSCALL(open, 3, ISGX_FILE, O_RDWR | O_CLOEXEC, 0);
                     ^~~~~~~~~~~~~~
In file included from sgx_framework.c:5:0:
sgx_framework.c:32:46: error: expected ‘)’ before ‘ISGX_FILE’
         SGX_DBG(DBG_E, "Cannot open device " ISGX_FILE ". "
                                              ^
pal_linux.h:340:24: note: in definition of macro ‘SGX_DBG’
             pal_printf(fmt);     \
                        ^~~
sgx_framework.c: In function ‘create_enclave’:
sgx_framework.c:178:12: error: variable ‘param’ has initializer but incomplete type
     struct sgx_enclave_create param = {
            ^~~~~~~~~~~~~~~~~~
sgx_framework.c:179:10: error: ‘struct sgx_enclave_create’ has no member named src’
         .src = (uint64_t)secs,
          ^~~
sgx_framework.c:179:16: warning: excess elements in struct initializer
         .src = (uint64_t)secs,
                ^
sgx_framework.c:179:16: note: (near initialization for ‘param’)
sgx_framework.c:178:31: error: storage size of ‘param’ isn’t known
     struct sgx_enclave_create param = {
                               ^~~~~
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18:0,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c:181:55: error: ‘SGX_IOC_ENCLAVE_CREATE’ undeclared (first use in this function); did you mean ‘DEFAULT_ENCLAVE_BASE’?
     int ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_CREATE, &param);
                                                       ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:174:26: note: in definition of macro ‘LOAD_ARGS_2’
     long __arg2 = (long)(a2); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:181:15: note: in expansion of macro ‘INLINE_SYSCALL’
     int ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_CREATE, &param);
               ^~~~~~~~~~~~~~
sgx_framework.c:178:31: warning: unused variable ‘param’ [-Wunused-variable]
     struct sgx_enclave_create param = {
                               ^~~~~
sgx_framework.c: In function ‘add_pages_to_enclave’:
sgx_framework.c:335:12: error: variable ‘param’ has initializer but incomplete type
     struct sgx_enclave_add_page param = {
            ^~~~~~~~~~~~~~~~~~~~
sgx_framework.c:336:10: error: ‘struct sgx_enclave_add_page’ has no member named ‘addr’
         .addr    = (uint64_t)addr,
          ^~~~
sgx_framework.c:336:20: warning: excess elements in struct initializer
         .addr    = (uint64_t)addr,
                    ^
sgx_framework.c:336:20: note: (near initialization for ‘param’)
sgx_framework.c:337:10: error: ‘struct sgx_enclave_add_page’ has no member named ‘src’
         .src     = (uint64_t)(user_addr ?: g_zero_pages),
          ^~~
sgx_framework.c:337:20: warning: excess elements in struct initializer
         .src     = (uint64_t)(user_addr ?: g_zero_pages),
                    ^
sgx_framework.c:337:20: note: (near initialization for ‘param’)
sgx_framework.c:338:10: error: ‘struct sgx_enclave_add_page’ has no member named ‘secinfo’
         .secinfo = (uint64_t)&secinfo,
          ^~~~~~~
sgx_framework.c:338:20: warning: excess elements in struct initializer
         .secinfo = (uint64_t)&secinfo,
                    ^
sgx_framework.c:338:20: note: (near initialization for ‘param’)
sgx_framework.c:339:10: error: ‘struct sgx_enclave_add_page’ has no member named ‘mrmask’
         .mrmask  = skip_eextend ? 0 : (uint16_t)-1,
          ^~~~~~
sgx_framework.c:339:20: warning: excess elements in struct initializer
         .mrmask  = skip_eextend ? 0 : (uint16_t)-1,
                    ^~~~~~~~~~~~
sgx_framework.c:339:20: note: (near initialization for ‘param’)
sgx_framework.c:335:33: error: storage size of ‘param’ isn’t known
     struct sgx_enclave_add_page param = {
                                 ^~~~~
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18:0,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c:344:55: error: ‘SGX_IOC_ENCLAVE_ADD_PAGE’ undeclared (first use in this function)
         ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_ADD_PAGE, &param);
                                                       ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:174:26: note: in definition of macro ‘LOAD_ARGS_2’
     long __arg2 = (long)(a2); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:344:15: note: in expansion of macro ‘INLINE_SYSCALL’
         ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_ADD_PAGE, &param);
               ^~~~~~~~~~~~~~
sgx_framework.c:335:33: warning: unused variable ‘param’ [-Wunused-variable]
     struct sgx_enclave_add_page param = {
                                 ^~~~~
sgx_framework.c: In function ‘init_enclave’:
sgx_framework.c:380:12: error: variable ‘param’ has initializer but incomplete type
     struct sgx_enclave_init param = {
            ^~~~~~~~~~~~~~~~
sgx_framework.c:382:10: error: ‘struct sgx_enclave_init’ has no member named ‘addr’
         .addr = enclave_valid_addr,
          ^~~~
sgx_framework.c:382:17: warning: excess elements in struct initializer
         .addr = enclave_valid_addr,
                 ^~~~~~~~~~~~~~~~~~
sgx_framework.c:382:17: note: (near initialization for ‘param’)
sgx_framework.c:384:10: error: ‘struct sgx_enclave_init’ has no member named ‘sigstruct’
         .sigstruct = (uint64_t)sigstruct,
          ^~~~~~~~~
sgx_framework.c:384:22: warning: excess elements in struct initializer
         .sigstruct = (uint64_t)sigstruct,
                      ^
sgx_framework.c:384:22: note: (near initialization for ‘param’)
sgx_framework.c:386:10: error: ‘struct sgx_enclave_init’ has no member named ‘einittoken’
         .einittoken = (uint64_t)token,
          ^~~~~~~~~~
sgx_framework.c:386:23: warning: excess elements in struct initializer
         .einittoken = (uint64_t)token,
                       ^
sgx_framework.c:386:23: note: (near initialization for ‘param’)
sgx_framework.c:380:29: error: storage size of ‘param’ isn’t known
     struct sgx_enclave_init param = {
                             ^~~~~
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18:0,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c:389:55: error: ‘SGX_IOC_ENCLAVE_INIT’ undeclared (first use in this function); did you mean ‘ATOMIC_FLAG_INIT’?
     int ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_INIT, &param);
                                                       ^
../../../include/arch/x86_64/Linux/sysdep-arch.h:174:26: note: in definition of macro ‘LOAD_ARGS_2’
     long __arg2 = (long)(a2); \
                          ^~
../../../include/arch/x86_64/Linux/sysdep-arch.h:134:9: note: in expansion of macro ‘LOAD_ARGS_3’
         LOAD_ARGS_##nr(args);                            \
         ^~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:144:50: note: in expansion of macro ‘INTERNAL_SYSCALL_NCS’
 #define INTERNAL_SYSCALL(name, err, nr, args...) INTERNAL_SYSCALL_NCS(__NR_##name, err, nr, ##args)
                                                  ^~~~~~~~~~~~~~~~~~~~
../../../include/arch/x86_64/Linux/sysdep-arch.h:120:43: note: in expansion of macro ‘INTERNAL_SYSCALL’
 #define INLINE_SYSCALL(name, nr, args...) INTERNAL_SYSCALL(name, , nr, args)
                                           ^~~~~~~~~~~~~~~~
sgx_framework.c:389:15: note: in expansion of macro ‘INLINE_SYSCALL’
     int ret = INLINE_SYSCALL(ioctl, 3, g_isgx_device, SGX_IOC_ENCLAVE_INIT, &param);
               ^~~~~~~~~~~~~~
sgx_framework.c:399:18: error: ‘SGX_INVALID_SIG_STRUCT’ undeclared (first use in this function); did you mean ‘SGX_LINKABLE_SIGNATURE’?
             case SGX_INVALID_SIG_STRUCT:
                  ^~~~~~~~~~~~~~~~~~~~~~
                  SGX_LINKABLE_SIGNATURE
sgx_framework.c:402:18: error: ‘SGX_INVALID_ATTRIBUTE’ undeclared (first use in this function); did you mean ‘SGX_INVALID_SIG_STRUCT’?
             case SGX_INVALID_ATTRIBUTE:
                  ^~~~~~~~~~~~~~~~~~~~~
                  SGX_INVALID_SIG_STRUCT
sgx_framework.c:405:18: error: ‘SGX_INVALID_MEASUREMENT’ undeclared (first use in this function); did you mean ‘SGX_INVALID_ATTRIBUTE’?
             case SGX_INVALID_MEASUREMENT:
                  ^~~~~~~~~~~~~~~~~~~~~~~
                  SGX_INVALID_ATTRIBUTE
sgx_framework.c:408:18: error: ‘SGX_INVALID_SIGNATURE’ undeclared (first use in this function); did you mean ‘SGX_INVALID_SIG_STRUCT’?
             case SGX_INVALID_SIGNATURE:
                  ^~~~~~~~~~~~~~~~~~~~~
                  SGX_INVALID_SIG_STRUCT
sgx_framework.c:411:18: error: ‘SGX_INVALID_LICENSE’ undeclared (first use in this function); did you mean ‘SGX_INVALID_SIGNATURE’?
             case SGX_INVALID_LICENSE:
                  ^~~~~~~~~~~~~~~~~~~
                  SGX_INVALID_SIGNATURE
sgx_framework.c:414:18: error: ‘SGX_INVALID_CPUSVN’ undeclared (first use in this function); did you mean ‘SGX_INVALID_LICENSE’?
             case SGX_INVALID_CPUSVN:
                  ^~~~~~~~~~~~~~~~~~
                  SGX_INVALID_LICENSE
sgx_framework.c:380:29: warning: unused variable ‘param’ [-Wunused-variable]
     struct sgx_enclave_init param = {
                             ^~~~~
Makefile:117: recipe for target 'sgx_framework.o' failed
make[3]: *** [sgx_framework.o] Error 1
make[3]: Leaving directory '/home/arushiarora/sgx/graphene/Pal/src/host/Linux-SGX'
Makefile:101: recipe for target 'host_lib_recurse' failed
make[2]: *** [host_lib_recurse] Error 2
make[2]: Leaving directory '/home/arushiarora/sgx/graphene/Pal/src'
Makefile:7: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/home/arushiarora/sgx/graphene/Pal'
Makefile:13: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by dimakuv** [comment]
It looks like you had an old version of Graphene and updated to the latest version in the same directory (which is perfectly valid). We recently changed the detection of the SGX driver in Graphene, which leads to the situation that you observe.

Please do `make SGX=1 distclean` to remove all remnants of the old built artifacts. Also, if this doesn't help, try explicit `rm -rf Pal/src/host/Linux-SGX/gsgx.h` (this remove the auto-generated `gsgx.h` file).

Then do the usualy `make SGX=1` again. Graphene should build cleanly now.

**Comment by noodlehair-git** [comment]
Thanks so much for your reply but none of the options seem to work for me. Is there anything else I could possibly do? There is also a ```Pal/src/host/Linux-SGX/gsgx.h.in```. Can it be something related to that?

**Comment by mkow** [comment]
Try `git clean -dfx` and then building again with `ISGX_DRIVER_PATH` set correctly. But be careful - this will restore the whole repository to a clean state and delete _all_ the files which are not part of the repository.

Such issues will hopefully be a history soon, we plan to migrate to Meson and have out-of-tree builds.

**Comment by noodlehair-git** [comment]
I rebooted and reinstalled graphene in a different directory this time. I get the second error again (in my original question). Shall I try installing an older version?

**Comment by dimakuv** [comment]
Are you sure you correctly set `ISGX_DRIVER_PATH`? It looks like the build doesn't see this env variable or it sees a wrong one...

**Comment by noodlehair-git** [comment]
I have set it as ```/home/sgx/linux-sgx-driver/``` i.e the path to the sgx driver c files. While installing graphene sgx driver (for me that is at ```/graphene/Pal/src/host/Linux-SGX/graphene-sgx-driver```) previously, I always gave this path (to my sgx driver c files) and it all used to work fine.

**Comment by noodlehair-git** [comment]
So I tried with your older version that I was using previously (I downloaded that sometime in Oct'20). I faced no issues while installing that. Also, you had ```sgx-driver``` then instead of ```graphene-sgx-driver``` (I'm not sure if anything here was changed except the name; Also this is before FSGSBASE changes).

**Comment by dimakuv** [comment]
Strange, it should work... I can't reproduce your error.

This is the example of how we build it in our automated builds in Jenkins: https://github.com/oscarlab/graphene/blob/master/.ci/lib/stage-build-sgx.jenkinsfile#L26

What is the version of your Linux SGX driver?

**Comment by noodlehair-git** [comment]
I used the 29th May'20 commit from https://github.com/intel/linux-sgx-driver/blob/master/README.md . 
I installed it in a custom directory ```/home/sgx/linux-sgx-driver```. Graphene and sgxsdk are also in the same ```sgx``` directory.

**Comment by dimakuv** [comment]
I didn't find a commit from 29th May 2020 there: https://github.com/intel/linux-sgx-driver/commits/master. I assume you're using commit 4e1dd1b13665cf70dce6336258512680573355b6 (from 30th of May).

Well, I can't reproduce the issue on my machine:
```
$ history
 2028  mkdir bla
 2029  cd bla/
 2030  git clone https://github.com/intel/linux-sgx-driver.git
 2031  cd linux-sgx-driver/
 2032  git checkout 4e1dd1b13665cf70dce6336258512680573355b6
 2033  make
 2034  cd ../
 2035  git clone https://github.com/oscarlab/graphene.git
 2036  cd graphene/
 2037  ISGX_DRIVER_PATH="$HOME/bla/linux-sgx-driver/" make SGX=1
```

Everything builds fine...

**Comment by noodlehair-git** [comment]
I was installing the ```graphene-sgx-driver``` first (before "making" graphene) . I followed your steps and it seems to work. Thanks.
